lightning/src/ln/peers/conduit.rs

```diff
@@ -33,13 +33,29 @@ pub(super) struct Decryptor {
 
 	pending_message_length: Option<usize>,
 	read_buffer: Option<Vec<u8>>,
+	poisoned: bool, // signal an error has occurred so None is returned on iteration after failure
 }
 
 impl Iterator for Decryptor {
-	type Item = Vec<u8>;
+	type Item = Result<Option<Vec<u8>>, String>;
 
 	fn next(&mut self) -> Option<Self::Item> {
-		self.decrypt_single_message(None)
+		if self.poisoned {
+			return None;
+		}
+
+		match self.decrypt_single_message(None) {
+			Ok(Some(result)) => {
+				Some(Ok(Some(result)))
+			},
+			Ok(None) => {
+				None
+			}
+			Err(e) => {
+				self.poisoned = true;
```

Comment:
I think there is a slightly bigger consideration here. Once this is poisoned, one option might be plain disconnecting the peer, because we will not be able to decrypt any future messages.

The other option would be temporarily caching the current nonce in the decrypt method, and if decryption fails, reverting the object's nonce property to the cached value.

Reply:
I looked through the code on master and it looks like it just disconnects unconditionally on decryption failure. Are you suggesting we just drain the current set of bytes (18 for header or len if we decrypted header) and try again?

Is this a common pattern for this type of communication decryption failure? I'm not very familiar with the different failure types and what should be recoverable and what should just give up.

Even if the behavior change is an improvement it may make sense to keep master functionality in the 494+ cleanup and add additional features later to minimize the behavior changes vs. refactor changes.

I hadn't realized this use case so I appreciate you starting the conversation.

Reply:
I think, upon reflection, that an immediate disconnect is the safest way to proceed. As long as now premature decryption attempts are made, there is really no reason for trying to salvage what should be unsalvageable.

Reply:
Resolved: Keeping the d/c on error behavior